Feature/binding sub propertyids (fixes #9210)

OlliTietavainenVaadin opened the pull request:
Fix for #9210 - allow use of nested properties with Binder when using the @PropertyId annotation.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/framework/9925)
<!-- Reviewable:end -->

OlliTietavainenVaadin committed:
Allow binding sub properties when using annotation @propertyId, fixes #9210

OlliTietavainenVaadin committed:
added missing @since tags

OlliTietavainenVaadin committed:
add missing serialVersionUIDs

OlliTietavainenVaadin commented:
FYI - trying to figure out the test failures. ClassesSerializableTest is simple to fix, BeanBinderTest not so much.

OlliTietavainenVaadin committed:
fixes BeanValidatorBinder not working with same NestedBeanPropertyDefinition naming as annotation-based scanning

tsuoanttila commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 241 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KuTru_Dte_pQCvsdGRR:-KuTru_Dte_pQCvsdGRS:b-7rxlpm) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L241)):*
> ```Java
>          * @param useLongFormName
>          *            use format grandparent.parent.property for name if true
>          * @since 8.y.x
> ```

See comment about empty since tags.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 486 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KtVa0M9SHWnATg60tDC:-KtVa0M9SHWnATg60tDD:b-d3fgxt) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L486)):*
> ```Java
>      *            filtering conditions for nested properties
>      * @return the bean property set, not <code>null</code>
>      * @since 8.x.y
> ```

Rather use empty since tag, than .x.y. Our build script will alert on the empty one and it will get filled, but this would go unnoticed if not handled in review step.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1101 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KtVaDcsiQ3HM2vF7321:-KtVaDcsiQ3HM2vF7322:bd5fe75) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/Binder.java#L1101)):*
> ```Java
>      * Class containing the constraints for filtering nested properties.
>      *
>      * @since 8.x.y
> ```

see comment about empty since

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1164 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KtVaGlWA_suhKpXMhQS:-KtVaGlWA_suhKpXMhQT:bd5fe75) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/Binder.java#L1164)):*
> ```Java
>      * @param scanNestedDefinitions
>      *            if true, scan for nested property definitions as well
>      * @since 8.x.y
> ```

see comment about empty since

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
since tags

OlliTietavainenVaadin commented:
Review status: 0 of 4 files reviewed at latest revision, 4 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 241 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KuTru_Dte_pQCvsdGRR:-KuY4Thieyiet-VC3IRW:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L241)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

See comment about empty since tags.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 486 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KtVa0M9SHWnATg60tDC:-KuY4UJkdOo_HqtKdPJn:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L486)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Rather use empty since tag, than .x.y. Our build script will alert on the empty one and it will get filled, but this would go unnoticed if not handled in review step.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1101 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KtVaDcsiQ3HM2vF7321:-KuY4VAbBFyxq87sxMw1:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/Binder.java#L1101)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

see comment about empty since
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1164 at r3](https://reviewable.io:443/reviews/vaadin/framework/9925#-KtVaGlWA_suhKpXMhQS:-KuY4Vgxde8Uib2kngmw:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/b10945b86063c6d4a6d61a5625400c65dd8cfa91/server/src/main/java/com/vaadin/data/Binder.java#L1164)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

see comment about empty since
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 2 of 4 files at r1.
Review status: 2 of 4 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 254 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRVT_MVwBMPdWO8T5F:-KvRVT_MVwBMPdWO8T5G:bek7mti) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L254)):*
> ```Java
>             return bean -> {
>                 Method readMethod = getDescriptor().getReadMethod();
>                 ValueProvider<T, ?> parentGetter = parent.getGetter();
> ```

Why this change? Should it also apply to getSetter() if it is needed here?

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 309 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRVf0Z9NdZ99xu3yNZ:-KvRVf0Z9NdZ99xu3yN_:blnu9fw) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L309)):*
> ```Java
> 
>     /**
>      * Key for identifying cached BeanPropertySet instances
> ```

End first sentence of javadoc with a period. Add since tag (even though not public, but as it warrants javadoc anyway).

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 399 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRf_wJHuRPdaL8O5yb:-KvRf_wJHuRPdaL8O5yc:b-w60hdp) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L399)):*
> ```Java
>     }
> 
>     private BeanPropertySet(Class<T> beanType, boolean checkNestedDefinitions,
> ```

(The mechanism could be made more extensible by opening up these methods, but that would require more javadoc and maybe more API design effort.)

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 409 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRiGkMsvnmxWvNfir6:-KvRiGkMsvnmxWvNfir7:b-92yule) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L409)):*
> ```Java
>     }
> 
>     private void findNestedDefinitions(
> ```

(not yet reviewed in detail)

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 98 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRb5XQdzdRi2n3Vnq2:-KvRb5XQdzdRi2n3Vnq3:bbnu19h) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L98)):*
> ```Java
> public class Binder<BEAN> implements Serializable {
> 
>     protected static final int MAX_PROPERTY_NESTING_DEPTH = 10;
> ```

If protected, this could have javadoc and a since tag. Does this belong here or in BeanPropertySet, or perhaps somewhere else?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1099 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRaWZ647dx5q6owW8c:-KvRaWZ647dx5q6owW8d:b4z7nv1) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L1099)):*
> ```Java
> 
>     /**
>      * Class containing the constraints for filtering nested properties.
> ```

Not a blocker, but somewhere (perhaps on the constructors of Binder rather than here) there could be information about how to customize filtering.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1104 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRaoQjCI5MwFnCxRa7:-KvRaoQjCI5MwFnCxRa8:b-3lzgws) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L1104)):*
> ```Java
>      *
>      */
>     protected static class PropertyFilterDefinition implements Serializable {
> ```

Does this class really belong to Binder, or should it perhaps be in BeanPropertySet which actually uses it?

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1146 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRaBMFVhpvHBDbG7wi:-KvRaBMFVhpvHBDbG7wj:byg1zo6) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L1146)):*
> ```Java
>     /**
>      * Creates a new binder that uses reflection based on the provided bean type
>      * to resolve bean properties. Allows scanning for nested properties as
> ```

How does this constructor allow scanning of nested properties? Should this just call the other constructor with suitable parameter values, or would that change the semantics regardless of the values?

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 403 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRcMmAEsbumlgpA9I0:-KvRcMmAEsbumlgpA9I1:bbtm6mq) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L403)):*
> ```Java
> 
>     }
> ```

no tests for any negative cases such as limit of nesting (circular dependencies etc.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
add test for binder with circular nesting properties

OlliTietavainenVaadin committed:
refactor nested property scanning code from Binder to BeanPropertySet

OlliTietavainenVaadin commented:
Review status: 1 of 4 files reviewed at latest revision, 9 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 254 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRVT_MVwBMPdWO8T5F:-KvVrcA6pNFRAYECiqjs:b-hnvhzb) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L254)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Why this change? Should it also apply to getSetter() if it is needed here?
</blockquote></details>

It was to make debugging a bit clearer, reverted

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 309 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRVf0Z9NdZ99xu3yNZ:-KvVrlor76Pq7tLSxWDf:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L309)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

End first sentence of javadoc with a period. Add since tag (even though not public, but as it warrants javadoc anyway).
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 399 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRf_wJHuRPdaL8O5yb:-KvVrohIwME8aRaP5l2t:b-z2krfm) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L399)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(The mechanism could be made more extensible by opening up these methods, but that would require more javadoc and maybe more API design effort.)
</blockquote></details>

Left as-is for now.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 409 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRiGkMsvnmxWvNfir6:-KvVrrtUeX33NlAOlO1D:b2nhll9) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L409)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(not yet reviewed in detail)
</blockquote></details>

Ok.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 98 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRb5XQdzdRi2n3Vnq2:-KvVrt5QKDrWDiYMSZiX:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L98)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

If protected, this could have javadoc and a since tag. Does this belong here or in BeanPropertySet, or perhaps somewhere else?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1099 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRaWZ647dx5q6owW8c:-KvVrw_nglB5S7-YFP4v:b87gbbf) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L1099)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Not a blocker, but somewhere (perhaps on the constructors of Binder rather than here) there could be information about how to customize filtering.
</blockquote></details>

Left as is for now.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1104 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRaoQjCI5MwFnCxRa7:-KvVrzQi5FTLnD7Y5EgD:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L1104)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Does this class really belong to Binder, or should it perhaps be in BeanPropertySet which actually uses it?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1146 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRaBMFVhpvHBDbG7wi:-KvVs-8oeGfvpoj2Jr8G:b-14opvm) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/main/java/com/vaadin/data/Binder.java#L1146)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

How does this constructor allow scanning of nested properties? Should this just call the other constructor with suitable parameter values, or would that change the semantics regardless of the values?
</blockquote></details>

It doesn't, that was a copy-paste error, removed the comment.

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 403 at r4](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvRcMmAEsbumlgpA9I0:-KvVs3JVFVAFAUR8nz2C:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/d62b3f5aa7126c021907482abcd493c2c7737483/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L403)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

no tests for any negative cases such as limit of nesting (circular dependencies etc.)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

hesara commented:
Reviewed 2 of 3 files at r5.
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 36 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaA_69oLtcm68_4Kdh:-KvaA_69oLtcm68_4Kdi:b91axwr) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L36)):*
> ```Java
> import java.util.stream.Stream;
> 
> import com.vaadin.data.BeanPropertySet.NestedBeanPropertyDefinition.PropertyFilterDefinition;
> ```

(feels a bit funny to import a class inside the same file)

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 235 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaApb1Q_2UW5khivRg:-KvaApb1Q_2UW5khivRh:b-cf7mid) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L235)):*
> ```Java
>         protected static class PropertyFilterDefinition
>                 implements Serializable {
>             private static final long serialVersionUID = 7035499395627976263L;
> ```

Actually the framework does not really use explicit serialVersionUID's as we don't manage compatibility between versions. Serialization works without these between cluster nodes or over a server restart, and a serialVersionUID just gives the impression we guarantee more than we really do.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 239 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaC21vKyl2iV9fnkch:-KvaC21vKyl2iV9fnkci:bffczou) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L239)):*
> ```Java
>             private List<String> ignorePackageNamesStartingWith;
> 
>             public PropertyFilterDefinition(int maxNestingDepth,
> ```

javadoc, including explaining the parameters

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 295 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBTXn1Lc2r-GslIOl:-KvaBTXn1Lc2r-GslIOm:b931oxt) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L295)):*
> ```Java
>          *            property descriptor
>          * @param useLongFormName
>          *            use format grandparent.parent.property for name if true
> ```

What would be the cases where one wants to use the long form, and what would be the cases where one doesn't?

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 368 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBCTjrAk5m64Tl2ME:-KvaBCTjrAk5m64Tl2MF:b-ao8hyl) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L368)):*
> ```Java
>      */
>     private static class InstanceKey implements Serializable {
>         private static final long serialVersionUID = -1363522555010923369L;
> ```

no need for serialVersionUID

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 469 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBrO5hp4MVa07IM2c:-KvaBrO5hp4MVa07IM2d:b-bdvgqk) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L469)):*
> ```Java
>             PropertyFilterDefinition filterCallback) {
>         if (depth >= filterCallback.getMaxNestingDepth()) {
>             return;
> ```

Should this just result in the problem being silently ignored, or an exception being thrown?

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 62 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-Kva8WVkVgvJp0el6fGC:-Kva8WVkVgvJp0el6fGD:b-od5wcv) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L62)):*
> ```Java
>     }
> 
>     public static class BindDeepNestingFieldsWithCircularStructure
> ```

(technically not circular but in practice tests the same case)

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 466 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBwCkrqIM35fzF6Yd:-KvaBwCkrqIM35fzF6Ye:bmrgklf) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L466)):*
> ```Java
>         Assert.assertNotNull(form.eighthLevelGrandchildName);
>         // only 10 levels of nesting properties are scanned by default
>         Assert.assertNull(form.distantGreatGrandchildName);
> ```

see the comment on nesting limit above

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
adds javadocs, small refactors

OlliTietavainenVaadin commented:
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 36 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaA_69oLtcm68_4Kdh:-KvakOtu08JvwEv7Q4rK:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L36)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(feels a bit funny to import a class inside the same file)
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 235 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaApb1Q_2UW5khivRg:-KvakPsp7JCcqgcEcHEz:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L235)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Actually the framework does not really use explicit serialVersionUID's as we don't manage compatibility between versions. Serialization works without these between cluster nodes or over a server restart, and a serialVersionUID just gives the impression we guarantee more than we really do.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 239 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaC21vKyl2iV9fnkch:-KvakQjnCz3402b0QA3a:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L239)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

javadoc, including explaining the parameters
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 295 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBTXn1Lc2r-GslIOl:-KvakRrRqy4je-540gBZ:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L295)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

What would be the cases where one wants to use the long form, and what would be the cases where one doesn't?
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 368 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBCTjrAk5m64Tl2ME:-KvakSv5HhqredM7bm9F:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L368)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

no need for serialVersionUID
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 469 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBrO5hp4MVa07IM2c:-KvakTirAvQtcR2H-WnD:bagsu2g) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L469)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

Should this just result in the problem being silently ignored, or an exception being thrown?
</blockquote></details>

An exception can't be thrown here, because we don't know when to stop recursing in the case of a property structure that contains (or its members contain) a member of its own type. See the test bindInstanceFields_circular() for an example of this.  It's not an exceptional or even uncommon state that loops exist within property trees, the recursion just needs to be terminated at some point.

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 62 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-Kva8WVkVgvJp0el6fGC:-KvamXRMJecWzd8Q5XYh:b58yjok) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L62)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

(technically not circular but in practice tests the same case)
</blockquote></details>

Couldn't come up with a better name.  "Nesting structure" isn't quite right either, since that also includes terminating nesting.

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 466 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBwCkrqIM35fzF6Yd:-Kvamq7QeSuSB0C8Tm2t:bh3kchu) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L466)):*
<details><summary><i>Previously, hesara (Henri Sara) wrote…</i></summary><blockquote>

see the comment on nesting limit above
</blockquote></details>

Seen. Default of ten levels of nesting should be enough for any reasonable use case, and for unreasonable use cases there's still the possibility of manually overriding the default max depth.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Review status: 3 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 469 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBrO5hp4MVa07IM2c:-KwdwkkY6QW75YtEB8Wk:bhmz03z) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L469)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

An exception can't be thrown here, because we don't know when to stop recursing in the case of a property structure that contains (or its members contain) a member of its own type. See the test bindInstanceFields_circular() for an example of this.  It's not an exceptional or even uncommon state that loops exist within property trees, the recursion just needs to be terminated at some point.
</blockquote></details>

Is there a way for the user to define the max nesting depth?

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 245 at r6](https://reviewable.io:443/reviews/vaadin/framework/9925#-Kwdw7HduwnCCnOMDMc_:-Kwdw7HduwnCCnOMDMca:bd1pxpb) ([raw file](https://github.com/vaadin/framework/blob/26d66aa136cdd47fd33465813812ce7761a2d73b/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L245)):*
> ```Java
>              *            sub-properties.
>              * @param ignorePackageNamesStartingWith
>              *            Ignore package names that start with this string, e.g.
> ```

replace "e.g." with "for example". No reason to use abbreviations in JavaDoc.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 304 at r6](https://reviewable.io:443/reviews/vaadin/framework/9925#-KwdwRQ4uzi4gIKpOcXm:-KwdwRQ4uzi4gIKpOcXn:b-ixzwng) ([raw file](https://github.com/vaadin/framework/blob/26d66aa136cdd47fd33465813812ce7761a2d73b/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L304)):*
> true
`{@code true}`

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 466 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBwCkrqIM35fzF6Yd:-Kwdvne3UHrMGHlaQgRJ:b-s22s70) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L466)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

Seen. Default of ten levels of nesting should be enough for any reasonable use case, and for unreasonable use cases there's still the possibility of manually overriding the default max depth.
</blockquote></details>

This and many other asserts in this test would benefit from an error message telling why the assertion failed. Especially mentioning the nesting depth out loud would make the developer who breaks this see why it broke.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
updated javadocs and unit test messages

OlliTietavainenVaadin commented:
Review status: 2 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 469 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBrO5hp4MVa07IM2c:-KwiucqtwqthOnOtQ-As:b-r0mq7u) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L469)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Is there a way for the user to define the max nesting depth?
</blockquote></details>

By extending Binder

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 245 at r6](https://reviewable.io:443/reviews/vaadin/framework/9925#-Kwdw7HduwnCCnOMDMc_:-KwiubWCLD0y8RYfmqS3:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26d66aa136cdd47fd33465813812ce7761a2d73b/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L245)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

replace "e.g." with "for example". No reason to use abbreviations in JavaDoc.
</blockquote></details>

Done.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 304 at r6](https://reviewable.io:443/reviews/vaadin/framework/9925#-KwdwRQ4uzi4gIKpOcXm:-KwiubptzeKxTVXMZDvc:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/26d66aa136cdd47fd33465813812ce7761a2d73b/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L304)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

> true
`{@code true}`
</blockquote></details>

Done.

---

*[server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java, line 466 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBwCkrqIM35fzF6Yd:-KwiuqfkXzFyyVloK060:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/test/java/com/vaadin/data/BinderInstanceFieldTest.java#L466)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

This and many other asserts in this test would benefit from an error message telling why the assertion failed. Especially mentioning the nesting depth out loud would make the developer who breaks this see why it broke.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Reviewed 1 of 2 files at r7.
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 469 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBrO5hp4MVa07IM2c:-Kwj34BRJWs-VQgjOZ7x:b-rf95h9) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L469)):*
<details><summary><i>Previously, OlliTietavainenVaadin (Olli Tietäväinen) wrote…</i></summary><blockquote>

By extending Binder
</blockquote></details>

Was there a test that demonstrates how to do it? If not, please add. Our own tests are one of the most important ways to prove that these things can be done and it should be easy to do. If it's not easy, then add at least a protected API or a constructor parameter to achieve it.

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1115 at r7](https://reviewable.io:443/reviews/vaadin/framework/9925#-Kwj3YmWXJAPa7k1bChZ:-Kwj3YmWXJAPa7k1bCh_:b-tmlt23) ([raw file](https://github.com/vaadin/framework/blob/5ca7f2558201c7671efb5ba5fc6409447bd8c803/server/src/main/java/com/vaadin/data/Binder.java#L1115)):*
> ```Java
>      *            the bean type to use, not <code>null</code>
>      * @param scanNestedDefinitions
>      *            if true, scan for nested property definitions as well
> ```

`{@code true}` seems I missed this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
added unit test for custom nesting level property scanning

OlliTietavainenVaadin committed:
fixed javadoc to use {@code }

OlliTietavainenVaadin commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[server/src/main/java/com/vaadin/data/BeanPropertySet.java, line 469 at r5](https://reviewable.io:443/reviews/vaadin/framework/9925#-KvaBrO5hp4MVa07IM2c:-Kxr2SuCBnQogGtT_Oqy:b-iw1p49) ([raw file](https://github.com/vaadin/framework/blob/4960e6d391c3334bb237ce9c8a7fb3c7d573dc6c/server/src/main/java/com/vaadin/data/BeanPropertySet.java#L469)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

Was there a test that demonstrates how to do it? If not, please add. Our own tests are one of the most important ways to prove that these things can be done and it should be easy to do. If it's not easy, then add at least a protected API or a constructor parameter to achieve it.
</blockquote></details>

Added test

---

*[server/src/main/java/com/vaadin/data/Binder.java, line 1115 at r7](https://reviewable.io:443/reviews/vaadin/framework/9925#-Kwj3YmWXJAPa7k1bChZ:-Kxr2UXmurOvpUn-jLti:b-896fix) ([raw file](https://github.com/vaadin/framework/blob/5ca7f2558201c7671efb5ba5fc6409447bd8c803/server/src/main/java/com/vaadin/data/Binder.java#L1115)):*
<details><summary><i>Previously, tsuoanttila (Teemu Suo-Anttila) wrote…</i></summary><blockquote>

`{@code true}` seems I missed this.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Needs merge from master, other than that <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925#-:-Kxr5e0lfjhFGpaCTwPA:bgjfgsy)*
<!-- Sent from Reviewable.io -->

OlliTietavainenVaadin committed:
merged

OlliTietavainenVaadin committed:
fixes BeanPropertySetTest

OlliTietavainenVaadin commented:
Merged

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925#-:-KxrLhnoZkllRSUPwWPS:b7t3kwp)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Reviewed 3 of 4 files at r9.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Reviewed 1 of 4 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/framework/9925)*
<!-- Sent from Reviewable.io -->

tsuoanttila commented:
Github says: CLA not signed. Did you commit from a new email at some point ?

OlliTietavainenVaadin commented:
WTF. No, email hasn't changed. Tried to sign it again, that didn't change anything.

OlliTietavainenVaadin commented:
Okay, now checks pass.
